main/src/library/BPlusTree.flix

```diff
@@ -0,0 +1,763 @@
+        ///
+        /// Not thread-safe.
+        ///
+        pub def rangeQueryWith(
+                f: k -> v -> Unit \ r0,
+                min: k,
+                max: k,
+                node: Node[k, v, r1]): Unit \ r0 + r1 with Order[k] =
+            let minLeaf = findLeafUnsafe(min, node);
+            let index = binarySearch(min, minLeaf->size, minLeaf->keys);
+            if(index < 0) {
+                let insertionPoint = toInsertionPoint(index);
+                if (insertionPoint <= minLeaf->size)
+                    traverseRight(f, insertionPoint, max, minLeaf)
+                    else ()
+            } else
+                traverseRight(f, index, max, minLeaf)
+
+        ///
+        /// Applies `f` to key-value pairs `(k, v)` starting at `node->keys[index]` until `node->keys[index] > max`.
+        ///
+        /// Uses the `next` pointer to traverse to the next node.
+        ///
+        def traverseRight(
+                f: k -> v -> Unit \ r0,
+                index: Int32,
+                max: k,
+                node: Node[k, v, r1]): Unit \ r0 + r1 with Order[k] =
+            if(index < node->size) {
+                let k = Array.get(index, node->keys);
+                if(k <= max) {
+                    f(k, Array.get(index, node->values));
+                    traverseRight(f, index + 1, max, node)
+                } else ()
+            } else match node->next {
+                case Some(next) => traverseRight(f, 0, max, next)
+                case None => ()
+            }
+
+        /////////////////////////////////////////////////////////////////////////////
+        // Helper methods                                                          //
+        /////////////////////////////////////////////////////////////////////////////
+
+        ///
+        /// Insert `x` at position `i` in array `arr` as though it currently has
+        /// `len` elements. Shifts elements at position `[i, .., len]` to `[i+1, .., len+1]`.
+        ///
+        def arrayInsert(x: a, i: Int32, len: Int32, arr: Array[a, r]): Unit \ r =
+            if (len == i)
+                Array.put(x, i, arr)
+            else
+                Array.copyInto(srcPos = i, dstPos = i + 1, len = len - i, src = arr, arr);
+                Array.put(x, i, arr)
+
+        ///
+        /// Returns `true` if the pointer of `o1` is equal to the pointer of `o2`.
+        ///
+        def ptrEqual(o1: v, o2: v): Bool =
+            // Unchecked_cast is here used to transform the values to Object.
+            // The purpose is to be able to give them to a Java class which performs pointer comparison.
+            let obj1 = unchecked_cast(o1 as Object);
+            let obj2 = unchecked_cast(o2 as Object);
+            // Unchecked_cast is here used to remove the IO effect from interacting with Java.
+            unchecked_cast(Objects.equals(obj1, obj2) as _ \ {})
```

Comment:
put `(Objects.equals(obj1, obj2): _ \ IO)` on the inside so we know that you are removing IO

(sorry about this stuff, hopefully we will improve it soon)